backend/FwHeadless/Controllers/MediaFileController.cs

```diff
@@ -0,0 +1,305 @@
+using FwHeadless.Models;
+using LexCore.Entities;
+using LexData;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Http.HttpResults;
+using Microsoft.Extensions.Options;
+using System.Security.Cryptography;
+using Microsoft.Net.Http.Headers;
+using System.Globalization;
+using System.Text.Json;
+using MimeMapping;
+
+namespace FwHeadless.Controllers;
+
+public static class MediaFileController
+{
+    public static async Task<Results<Ok<FileListing>, NotFound>> ListFiles(
+        Guid projectId,
+        IOptions<FwHeadlessConfig> config,
+        LexBoxDbContext lexBoxDb,
+        string relativePath = "")
+    {
+        var project = await lexBoxDb.Projects.FindAsync(projectId);
+        if (project is null) return TypedResults.NotFound();
+        var projectFolder = config.Value.GetProjectFolder(project.Code, projectId);
+        var path = Path.Join(projectFolder, relativePath); // Do NOT use Path.Combine here
+        var files =
+            Directory.EnumerateFiles(path, "*", new EnumerationOptions() { RecurseSubdirectories = true })
+                .Select(file => Path.GetRelativePath(projectFolder, file))
+                .ToArray();
+        return TypedResults.Ok(new FileListing(files));
+    }
+
+    public static async Task<Results<PhysicalFileHttpResult, NotFound>> GetFile(
+        Guid fileId,
+        IOptions<FwHeadlessConfig> config,
+        LexBoxDbContext lexBoxDb)
+    {
+        var mediaFile = await lexBoxDb.Files.FindAsync(fileId);
+        if (mediaFile is null) return TypedResults.NotFound();
+        var projectId = mediaFile.ProjectId;
+        var project = await lexBoxDb.Projects.FindAsync(projectId);
+        if (project is null) return TypedResults.NotFound();
+        var projectFolder = config.Value.GetProjectFolder(project.Code, projectId);
+        var filePath = Path.Join(projectFolder, mediaFile.Filename);
+        mediaFile.InitializeMetadataIfNeeded(filePath);
+        var contentType = mediaFile.Metadata.MimeType;
+        if (contentType is null)
+        {
+            contentType = MimeUtility.GetMimeMapping(filePath);
+            mediaFile.Metadata.MimeType = contentType;
+            await lexBoxDb.SaveChangesAsync();
+        }
+        if (await AddEntityTagMetadataIfNotPresent(mediaFile, filePath)) await lexBoxDb.SaveChangesAsync();
```

Comment:
this and the previous block should be done all at once, we shouldn't be saving changes twice if the data is missing.

Reply:
Done.